proposals/push-constants.md

```diff
@@ -0,0 +1,74 @@
+
+
+## Variable and Value Declarations
+
+| Declaration | Mutability | Scope | Effective-value-type | Initializer Support | Initializer Expression | Part of Resource Interface |
+| --- | --- | --- | --- | --- | --- | --- |
+| `var<immediate_data>` | Immutable | [Module](https://www.w3.org/TR/WGSL/#module-scope) | [Concrete](https://www.w3.org/TR/WGSL/#type-concrete) [constructible](https://www.w3.org/TR/WGSL/#constructible) [host-shareable](https://www.w3.org/TR/WGSL/#host-shareable) excludes [array types](https://www.w3.org/TR/WGSL/#array-types) and [structure types](https://www.w3.org/TR/WGSL/#struct-types) contains array members | Disallowed	| | Yes. [uniform buffer](https://www.w3.org/TR/WGSL/#uniform-buffer) |
+
+NOTE: Each [entry point](https://www.w3.org/TR/WGSL/#entry-point) can statically use at most one immediate data variable.
+
+# API
+
+## Limits
+
+One new limits:
+
+| Limit name | Description | Type | Limit class | Default |
+| --- | --- | --- | --- | --- |
+| immediateDataRangeMaxByteSize | The maximum bytes allowed value for the immediateDataRangeMaxSize | [GPUSize32](https://www.w3.org/TR/webgpu/#typedefdef-gpusize32) | [maximum](https://www.w3.org/TR/webgpu/#limit-class-maximum) | 64 |
+
+NOTE: 64 bytes is the sizeof(mat4x4).
+
+## Pipeline Layouts
+One new member in `GPUPipelineLayoutDescriptor`.
+
+```javascript
+dictionary GPUPipelineLayoutDescriptor
+         : GPUObjectDescriptorBase {
+    required sequence<GPUBindGroupLayout> bindGroupLayouts;
+    uint32_t immediateDataRangeByteSize = 0;
+};
+```
+`immediateDataRangeByteSize`: Size of immediate data range used in pipeline, type is bytes.
+
+NOTE: `immediateDataRangeByteSize` = sizeof(variables) + sizeof(paddings). Follow [ Aligment rules ](https://www.w3.org/TR/WGSL/#alignment-and-size) in wgsl spec.
+
+NOTE: two pipeline layouts are defined to be “compatible for immediate data” if they were created with identical immediate data byte size. It means immediate data values can share between pipeline layouts that are compatible for immediate data.
+
+NOTE: Immediate data range follow [out-of-bounds access](https://www.w3.org/TR/WGSL/#out-of-bounds-access) rules in wgsl spec.
+
+## GPUCommandEncoder
+
+Four new functions in `GPUCommandEncoder`.
+
+```javascript
+interface mixin GPUBindingCommandsMixin {
+        void setImmeidateDataU32(uint32_t offset, uint32_t value);
+        void setImmediateDataI32(uint32_t offset, int32_t value);
+        void setImmediateDataF32(uin32_t offset, float32_t value);
+        void setImmediateDataRange(uint32_t rangeOffset, AllowSharedBufferSource data, optional dataOffset, optional size);
```

Comment:
Do we really need to add four new APIs here? Or is the last one, `void setImmediateDataRange` which takes the  `AllowSharedBufferSource` sufficient?

Reply:
Thanks for reviewing! Considering that purpose of these APIs are 'easy to use' , I'm Ok to remove them.

Reply:
Could we pass the type as an argument and allow `setImmediate` to take a double similar to overridable constants? The concern is specifying the type in the method signature is not something used anywhere else in WebGPU. It seems inconsistent with the rest of the API.

Reply:
The reason why it's nice to have a method per type is that it makes it very explicit to the developer what gets written in terms of bit pattern. If we have a single `setImmediate(slot, Number)` then either we'd support only `f32` (and aggregates of) in immediates, or we'd have to use reflection on `draw` to know how to convert the bunch of `Number` we have to actual bit patterns.

Agreed though that the IDL for these methods should use the Javacript types (Number and typedefs in WebGPU's IDL) instead of C types ^^

Reply:
If not for the fact that TypedArrays have a really high allocation overhead (at least in V8) I would say
`setImmediateData(0, new Uint32Array([123]))` is really not that much worse than
`setImmediateU32(0, 123)`.
(Avoiding allocation overhead it ends up looking more like
```js
myTempU32[0] = 123;
setImmediateData(0, myTempU32);
```
)
I have never checked if this is just a V8 problem, but if it is, I think the need for the overloads is not that strong...

Reply:
BTW the lack of f16 in the WGSL definition doesn't make it useless to have a setImmediateF16 overload, you can still extract the value easily with `unpack2x16float`.
Though often you'd need to set two F16s at which point
```js
setImmediateF16(0, 10);
setImmediateF16(2, 20);
```
is really no better than
```js
setImmediateData(0, new Float16Array([10, 20]));
```

(Note Float16Array appears to be coming to JS soonish: https://github.com/tc39/proposal-float16array
https://crbug.com/42203953)

Reply:
Oh I was thinking like `setImmediate(slot, Number, type)` @Kangz where `type` is "int", "float", "uint", "fp16"

But would also agree with @kainino0x that `setImmediateData(0, new Uint32Array([123]))` and `setImmediateU32(0, 123)` are quite similar, at least semantically.

The other plus side to not having the overloads which I just thought of is it would encourage the developer to set all the immediate data at once. As opposed to issuing 16 = 64 / 4 calls to `setImmediateDataType` in the case all 64 bytes were used. So maybe that outweighs the ArrayBuffer creation and, as mentioned, the buffer could be reused so the creation cost is only on load

Reply:
To reinforce @kainino0x's point: TypedArray allocation overhead is really high in _all_ browsers, from what I've observed in the past. I think V8 is a little worse than others, but not by leaps and bounds.

As such I think the `setImmediate*32()` variants do have value.

Reply:
It would be good to have a performance benchmark or some comparison illustrating the need for introducing these overloads, before they are added to the specification and can not be removed without causing a breaking change.

Reply:
Agreed. Here's a benchmark of a f32-to-u32 reinterpretation, a common math operation one would do with TypedArray. https://jsbench.me/2vlxcmotny/1

| Case (run on M1 Pro)                    | Chrome Stable 125.0.6422.142 | Safari TP 195 (19619.1.15.0.1) | Firefox Nightly 128.0a1 |
|-----------------------------------------|------------------------------|--------------------------------|-------------------------|
| allocating_array_constructor | 3.2M ops/s ± 1.71%           | 5.7M ops/s ± 0.65%             | 3.3M ops/s ± 1.57%      |
| allocating_size_constructor  | 3.5M ops/s ± 1.95%           | 7M ops/s ± 0.64%               | 4M ops/s ± 1.55%        |
| allocating_wrappers_only     | 17M ops/s ± 2.1%             | 24M ops/s ± 0.45%              | 13M ops/s ± 3.42%       |
| nonallocating                | 216M ops/s ± 2.63%           | 1.4B ops/s ± 0.32%             | 1.6B ops/s ± 0.36%      |

That shows exactly what we expect, but I also tried a version which tries to defeat any const folding optimizations that would mess up the results, and it showed similar results. https://jsbench.me/2vlxcmotny/2

| Case (run on M1 Pro)                       | Chrome Stable 125.0.6422.142 | Safari TP 195 (19619.1.15.0.1) | Firefox Nightly 128.0a1 |
|--------------------------------------------|------------------------------|--------------------------------|-------------------------|
| 1M allocating_array_constructor | 4.2 ops/s ± 0.39%            | 5.5 ops/s ± 1.23%              | 3.2 ops/s ± 1.99%       |
| 1M allocating_size_constructor  | 4.4 ops/s ± 1.29%            | 6.8 ops/s ± 0.78%              | 3.8 ops/s ± 1.64%       |
| 1M allocating_wrappers_only     | 24 ops/s ± 0.33%             | 22 ops/s ± 0.48%               | 13 ops/s ± 3.73%        |
| 1M nonallocating                | 165 ops/s ± 1.5%             | 222 ops/s ± 0.3%               | 247 ops/s ± 0.59%       |